app/main.py

```diff
@@ -1,214 +1,51 @@
-        "power": 40,
-        "hp": 70,
-        "armour": [
-            {
-                "part": "breastplate",
-                "protection": 25,
-            }
-        ],
-        "weapon": {
-            "name": "Sword",
-            "power": 45
-        },
-        "potion": {
-            "name": "Blessing",
-            "effect": {
-                "hp": +10,
-                "power": +5,
-            }
-        }
-    }
-}
+from app.data.source_data import KNIGHTS
 
+from app.knight import Knight
+from app.ammunition.armour import Armour
+from app.ammunition.potion import Potion
+from app.ammunition.weapon import Weapon
 
-def battle(knightsConfig):
-    # BATTLE PREPARATIONS:
 
-    # lancelot
-    lancelot = knightsConfig["lancelot"]
+def create_knights(knights: dict) -> {Knight}:
+    knights_dict = {}
 
-    # apply armour
-    lancelot["protection"] = 0
-    for a in lancelot["armour"]:
-        lancelot["protection"] += a["protection"]
+    for params in knights.values():
+        name = params["name"]
+        power = params["power"]
+        hp = params["hp"]
 
-    # apply weapon
-    lancelot["power"] += lancelot["weapon"]["power"]
+        armour = [Armour(arm["part"], arm["protection"])
+                  for arm in params["armour"]]
 
-    # apply potion if exist
-    if lancelot["potion"] is not None:
-        if "power" in lancelot["potion"]["effect"]:
-            lancelot["power"] += lancelot["potion"]["effect"]["power"]
+        weapon = Weapon(params["weapon"]["name"],
+                        params["weapon"]["power"])
+        try:
+            potion = Potion(params["potion"]["name"],
+                            params["potion"]["effect"])
+        except TypeError:
+            potion = None
 
-        if "protection" in lancelot["potion"]["effect"]:
-            lancelot["protection"] += lancelot["potion"]["effect"]["protection"]
+        knights_dict[name] = Knight(name, power, hp, armour, weapon, potion)
 
-        if "hp" in lancelot["potion"]["effect"]:
-            lancelot["hp"] += lancelot["potion"]["effect"]["hp"]
+    return knights_dict
 
-    # arthur
-    arthur = knightsConfig["arthur"]
 
-    # apply armour
-    arthur["protection"] = 0
-    for a in arthur["armour"]:
-        arthur["protection"] += a["protection"]
+def battle(knights_config: dict) -> {}:
+    knights = create_knights(knights_config)
+    [knight.prepare_for_battle() for knight in knights.values()]
+    lancelot = knights["Lancelot"]
+    mordred = knights["Mordred"]
+    arthur = knights["Artur"]
+    red_knight = knights["Red Knight"]
 
-    # apply weapon
-    arthur["power"] += arthur["weapon"]["power"]
+    lancelot.hit(mordred)
+    mordred.hit(lancelot)
 
-    # apply potion if exist
-    if arthur["potion"] is not None:
-        if "power" in arthur["potion"]["effect"]:
-            arthur["power"] += arthur["potion"]["effect"]["power"]
+    arthur.hit(red_knight)
+    red_knight.hit(arthur)
```

Comment:
not sure about all this being so hard-coded here. what about if we wanted to see eg Red Knight vs Mordred? maybe this can be revised a bit

Reply:
done